q2_annotate/busco/utils.py

```diff
@@ -323,3 +402,103 @@ def _process_busco_results(results, sample_id, mag_id, file_name, additional_met
     }
 
     return results
+
+
+def _get_mag_lengths(bins: Union[MultiMAGSequencesDirFmt, MAGSequencesDirFmt]):
+    lengths = {}
+    if isinstance(bins, MultiMAGSequencesDirFmt):
+        for sample, mags in bins.sample_dict().items():
+            for mag_id, mag_fp in mags.items():
+                seq = skbio.io.read(mag_fp, format="fasta")
+                lengths[mag_id] = sum([len(s) for s in seq])
+        return pd.Series(lengths, name="length")
+    else:
+        for mag_id, mag_fp in bins.feature_dict().items():
+            seq = skbio.io.read(mag_fp, format="fasta")
+            lengths[mag_id] = sum([len(s) for s in seq])
+        return pd.Series(lengths, name="length")
+
+
+def _filter_unbinned_for_partition(unbinned_contigs, mag_partition, _filter_contigs):
+    """
+    Filters the unbinned contigs to match the sample IDs in a MAG partition.
+
+    Args:
+        unbinned_contigs (ContigSequencesDirFmt): The full unbinned contigs.
+        mag_partition (MultiMAGSequencesDirFmt): One partition of MAGs.
+        _filter_contigs (Action): QIIME 2 action to filter contigs.
+
+    Returns:
+        ContigSequencesDirFmt: Filtered unbinned contigs matching the partition samples.
+    """
+    sample_ids = list(mag_partition.view(MultiMAGSequencesDirFmt).sample_dict().keys())
+    metadata = Metadata(pd.DataFrame(index=pd.Index(sample_ids, name="ID")))
+    id_list = ", ".join([f"'{sid}'" for sid in sample_ids])
+    where = f"ID IN ({id_list})"
+    (filtered_unbinned,) = _filter_contigs(
+        contigs=unbinned_contigs, metadata=metadata, where=where
+    )
+    return filtered_unbinned
+
+
+def _get_fasta_files_from_dir(directory: Path) -> list:
+    # Only match FASTA extensions starting with '.fa' or '.fna'
+    return [f for f in directory.glob("*") if f.suffix in {".fa", ".fasta"}]
+
+
+def _count_contigs(file_paths: List[Path]) -> int:
+    """
+    Count the number of DNA sequences across a list of FASTA files.
+
+    Parameters
+    ----------
+    file_paths : list of Path
+        List of FASTA file paths (.fa, .fasta, .fna).
+
+    Returns
+    -------
+    int
+        Total number of sequences across all files.
+    """
+    total_sequences = 0
+
+    for fp in file_paths:
+        total_sequences += sum(
+            1 for _ in skbio.io.read(str(fp), format="fasta", constructor=skbio.DNA)
+        )
+
+    return total_sequences
+
+
+def _calculate_unbinned_percentage(
```

Comment:
Please add type annotation and fix the docstring - it says that the method expects DirFmts but in fact you are passing paths to it. Also, please follow the docstring format convention we are using in other places of this repository.